api-consistency.md

```diff
@@ -0,0 +1,169 @@
+API consistency guidelines
+==========================
+
+Preupgrade Assistant provides two separate APIs for writing modules:
+Python API and Bash API.  These languages are very different, so are
+common patterns and methods of solving programming problems.
+
+Yet, we want to keep both APIs retain similar "look & feel" so that
+module developers and maintainers that will need to switch get as little
+surprise as possible.
+
+In order to help developers of both APIs prevent diverging the behaviors
+too much, this document attempts to set interface guidelines that make
+maximum use of both language's common features.
+
+
+Functions
+---------
+
+ *  Python: all functions live under `preupg` module:
+
+        preupg.myfun()
```

Comment:
I don't know if I understand your intentions here. We have a python package named _'preupg'_ now. You would create new module _'preupg'_ (i.e. filename _preupg.py_) within this package and use it as a python API for Preupgrade Assistant modules? That doesn't look that clean or intuitive to me.

Well, currently the python module, serving as an API for Preupgrade Assistant modules, is named _'script_api'_ and it isn't the best name either. We can change the name, that's not a problem, however we need to have it thought through. I vote for something more selfexplanatory like 'modules_api.py'. Having the name _common.sh_ for shell API is something to look at too. It can be renamed to _modules_api.sh_ for example, but that's too much off topic.

If we agree on the _preupg_ prefix, we can have the same function name in both python and shell API, i.e. _preupg_myfun_, instead of just _myfun_ as it is in the APIs now. The same with the variables. It can have name _PREUPG_MYVAR_ in both APIs.

However I don't think that the prefix is necessary. It  would compromise clarity IMHO. In python it would be like:
`from preupg import modules_api`
`modules_api.preupg_log_risk(PREUPG_RISK_LOW)`
I would prefer the nonprefix version:
`from preupg import modules_api`
`modules_api.log_risk(RISK_LOW)`